spec/lib/data_hygiene/title_fetcher_spec.rb

```diff
@@ -0,0 +1,205 @@
+require 'rails_helper'
+require 'data_hygiene/title_fetcher'
+
+RSpec.describe DataHygiene::TitleFetcher do
+  let(:client) { double('GovDelivery::Client') }
+  let(:logger) { double(:logger, info: true, warn: true) }
+  subject { described_class.new(client: client, logger: logger) }
+
+  describe '#update_title' do
+    context "when we don't have a name from GovDelivery" do
+      let!(:subscriber_list) { create(:subscriber_list, gov_delivery_id: 'ABC_123', title: nil) }
+
+      it 'does not update the subscriber list' do
+        expect { subject.send(:update_title, subscriber_list, nil) }.to_not change { subscriber_list.title }
+        expect { subject.send(:update_title, subscriber_list, '') }.to_not change { subscriber_list.title }
+        expect(subject.stats).to eq({ not_found: ['ABC_123', 'ABC_123'] })
+      end
+    end
+
+    context 'when we have found a name from GovDelivery' do
+      context "when the subscriber list's title is missing" do
+        shared_examples 'updates the title' do |initial_title|
+          let(:new_title) { 'Education' }
+          let!(:subscriber_list) { create(:subscriber_list, gov_delivery_id: 'ABC_123', title: initial_title) }
+
+          it "updates the subscriber list's title" do
+            expect { subject.send(:update_title, subscriber_list, new_title) }
+              .to change { subscriber_list.title }
+              .from(initial_title)
+              .to(new_title)
+          end
+
+          it 'logs the change' do
+            expect(logger).to receive(:info).with("ABC_123: title updated to match GovDelivery topic")
+            subject.send(:update_title, subscriber_list, new_title)
+            expect(subject.stats).to eq({ updated: ['ABC_123'] })
+          end
+        end
+
+        shared_examples 'logs when saving fails' do |initial_title|
+          let!(:subscriber_list) { create(:subscriber_list, gov_delivery_id: 'ABC_123', title: initial_title) }
+          before do
+            allow(subscriber_list).to receive(:save).and_return(false)
+          end
+
+          it 'logs the failure' do
+            expect(logger).to receive(:warn).with("ABC_123: failed to update title")
+            subject.send(:update_title, subscriber_list, 'Education')
+            expect(subject.stats).to eq({ update_failed: ['ABC_123'] })
+          end
+        end
+
+        it_behaves_like 'updates the title', nil
+        it_behaves_like 'updates the title', ''
+
+        it_behaves_like 'logs when saving fails', nil
+        it_behaves_like 'logs when saving fails', ''
```

Comment:
I would potentially argue that when we only have 1 test in a shared_example it is easier to follow by just have a duplicated test. :)